builder/beacon_client.go

```diff
@@ -51,14 +62,52 @@ func (b *BeaconClient) isValidator(pubkey PubkeyHex) bool {
 }
 
 func (b *BeaconClient) getProposerForNextSlot(requestedSlot uint64) (PubkeyHex, error) {
-	/* Only returns proposer if requestedSlot is currentSlot + 1, would be a race otherwise */
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
-	if b.currentSlot+1 != requestedSlot {
-		return PubkeyHex(""), errors.New("slot out of sync")
+	if (requestedSlot / b.slotsInEpoch) > b.currentEpoch {
+		b.currEpochProposerMap = b.nextEpochProposerMap
+	}
+
+	nextSlotProposer, found := b.currEpochProposerMap[requestedSlot]
+	if !found {
+		log.Error("inconsistent proposer mapping", "requestSlot", requestedSlot, "slotProposerMap", b.currEpochProposerMap)
+		return PubkeyHex(""), errors.New("inconsistent proposer mapping")
+	}
+	return nextSlotProposer, nil
+}
+
+func (b *BeaconClient) updateValidatorsMap() error {
+	// Start regular slot updates
+	slotsPerEpoch := b.slotsInEpoch
+	durationPerSlot := time.Duration(b.secondsInSlot) * time.Second
+	durationPerEpoch := durationPerSlot * time.Duration(slotsPerEpoch)
+	// Every half epoch request validators map
+	timer := time.NewTicker(durationPerEpoch / 2)
+	for {
+		select {
+		case <-timer.C:
```

Comment:
Does the ticker tick straight away? To trigger it right away (and not wait) you could set it to zero initially.
Or factor out the update routine to a function and call it before the ticker loop for initial setup (and it's okay to wait if it fails to connect).